modules/core/core-content/src/main/java/com/enonic/xp/core/impl/project/UpdateProjectRolesCommand.java

```diff
@@ -0,0 +1,193 @@
+        }
+
+        if ( !addedMembers.isEmpty() )
+        {
+            doAddRoleMembers( addedMembers, newProject.getName(), projectRole );
+        }
+        if ( !removedMembers.isEmpty() )
+        {
+            doRemoveRoleMembers( removedMembers, newProject.getName(), projectRole );
+        }
+    }
+
+    private void doAddRoleMembers( final Set<PrincipalKey> principalKeys, final ProjectName projectName, final ProjectRoles projectRole )
+    {
+        principalKeys.forEach( principalKey -> {
+            final PrincipalRelationship rel = PrincipalRelationship.
+                from( projectRole.getRoleKey( projectName ) ).
+                to( principalKey );
+
+            securityService.addRelationship( rel );
+        } );
+    }
+
+    private void doRemoveRoleMembers( final Set<PrincipalKey> principalKeys, final ProjectName projectName, final ProjectRoles projectRole )
+    {
+        principalKeys.
+            forEach( principalKey -> {
+
+                final PrincipalRelationship rel = PrincipalRelationship.
+                    from( projectRole.getRoleKey( projectName ) ).
+                    to( principalKey );
+
+                securityService.removeRelationship( rel );
+            } );
+    }
+
+    private Set<PrincipalKey> doGetAddedMembers( final Project newProject, final ProjectPermissionsLevel projectPermissionsLevel,
+                                                 final ProjectRoles projectRole )
+    {
+
+        final PrincipalKey roleKey = securityService.getRole( projectRole.getRoleKey( newProject.getName() ) ).
+            map( Role::getKey ).
+            orElse( null );
+
+        if ( roleKey == null )
+        {
+            return Set.of();
+        }
+
+        final PrincipalRelationships relationships = this.securityService.getRelationships( roleKey );
+        final List<PrincipalKey> members = relationships.stream().map( PrincipalRelationship::getTo ).collect( toList() );
+
+        return newProject.getPermissions().getPermission( projectPermissionsLevel ).
+            stream().
+            filter( principalKey -> !members.contains( principalKey ) ).
+            collect( Collectors.toSet() );
+    }
+
+    private Set<PrincipalKey> doGetRemovedMembers( final ProjectPermissions oldProjectPermissions,
+                                                   final ProjectPermissions newProjectPermissions,
+                                                   final ProjectPermissionsLevel projectPermissionsLevel )
+    {
+        return oldProjectPermissions.getPermission( projectPermissionsLevel ).
+            stream().
+            filter( principalKey -> !newProjectPermissions.getPermission( projectPermissionsLevel ).contains( principalKey ) ).
+            collect( Collectors.toSet() );
+    }
+
+    public static final class Builder
+        extends AbstractProjectRolesCommand.Builder<Builder>
+    {
+        private Project oldProject;
```

Comment:
I suggest 3 fields
projectName
oldMembers
newMembers
this way it won't be ambiguous.